packages/apps/spirekey-example/src/spirekey.ts

```diff
@@ -0,0 +1,167 @@
+import { IUnsignedCommand } from '@kadena/client';
+
+export class KadenaSpireKey {
+  private _returnUrl: string;
+  private _spireKeyHostname: string;
+  private _user: IUser | null = null;
+  private _transactions: Record<string, IUnsignedCommand> = {};
+
+  constructor(spireKeyHostname: string, returnUrl: string) {
+    this._spireKeyHostname = spireKeyHostname;
+    this._returnUrl = returnUrl;
+    this._loadFromLocalStorage();
+    this.update(window.location);
+  }
+
+  get isLoggedIn() {
+    return this._user !== null;
+  }
+
+  get user() {
+    return this._user;
+  }
+
+  get transactions() {
+    return this._transactions;
+  }
+
+  update(location: Location) {
+    const searchParams = new URLSearchParams(location.search);
+
+    if (searchParams.has('user')) {
+      const userSearch = searchParams.get('user');
+      if (userSearch && userSearch?.length > 0) {
+        // decode base64
+        this._user = JSON.parse(decodeBase64(userSearch));
+        console.log('retrieved user from querystring parameters', this._user);
+      }
+    }
+
+    if (searchParams.has('transaction')) {
+      const transactionSearch = searchParams.get('transaction');
+      if (transactionSearch && transactionSearch?.length > 0) {
+        const transaction: IUnsignedCommand = JSON.parse(
+          decodeBase64(transactionSearch),
+        );
+        console.log(
+          'retrieved transaction from querystring parameters',
+          JSON.stringify(transaction, null, 2),
+        );
+
+        this._transactions[transaction.hash] = transaction;
+      }
+    }
+    this._saveToLocalStorage();
+    // clear querystring parameters
+    history.pushState({}, '', location.pathname);
+  }
+
+  login() {
+    location.href = `${this._spireKeyHostname}/login?returnUrl=${this._returnUrl}`;
+  }
+
+  logout() {
+    this._clearLocalStorage();
+    location.href = new URL(this._returnUrl).origin;
+  }
+
+  sign(tx: IUnsignedCommand) {
+    const query: ISignRequestQuerystring = {
+      transaction: encodeBase64(JSON.stringify(tx)),
+      returnUrl: this._returnUrl,
+      optimistic: false,
+    };
+    const queryString = new URLSearchParams(query as any).toString();
+    location.href = `${this._spireKeyHostname}/sign?${queryString}`;
+  }
+
+  private _saveToLocalStorage() {
+    sessionStorage.setItem('user', JSON.stringify(this._user));
```

Comment:
u say save to localStorage but you store to session storage?
- session storage is cleared when closing browser / tab ( do we want that? )
- session storage can be accessed by other scripts... (risk)

Reply:
LocalStorage can be accessed by other scripts as well, right?